File: frontend/src/api/player/internal/admin/UpdateShopItemInternalApi.ts
import { z } from "zod";
import { APIMessage } from "../../../../system/api/APIMessage.js";

/** Mirrors backend internal API player/api/internal/admin/UpdateShopItemInternalApi.scala; not registered as a public frontend route. */
export class UpdateShopItemInternalAPI extends APIMessage<unknown> {
  readonly itemId: unknown;
  readonly name: unknown;
  readonly description: unknown;
  readonly price: unknown;
  readonly currency: unknown;
  readonly catalogIndex: unknown;
  readonly active: unknown;
  readonly sortOrder: unknown;
  constructor(itemId: unknown, name: unknown, description: unknown, price: unknown, currency: unknown, catalogIndex: unknown, active: unknown, sortOrder: unknown) {
    super();
    this.itemId = itemId;
    this.name = name;
    this.description = description;
    this.price = price;
    this.currency = currency;
    this.catalogIndex = catalogIndex;
    this.active = active;
    this.sortOrder = sortOrder;
  }

  override get responseSchema() {
    return z.unknown();
  }
}

File: backend/microservice/src/player/api/internal/admin/UpdateShopItemInternalApi.scala
package microservice.player.api.internal.admin

import cats.data.EitherT
import cats.effect.IO
import java.sql.Connection
import microservice.infrastructure.api.{APIMessage, PlanSteps}
import microservice.infrastructure.http.HttpError
import microservice.player.objects.shop.ShopItem
import microservice.player.support.shop.ShopCatalogSupport

/** 模块间 API：更新商店商品；由 admin HTTP API 调用，不挂路由。 */
final case class UpdateShopItemInternalAPIMessage(
  itemId: String,
  name: String,
  description: String,
  price: Int,
  currency: String,
  catalogIndex: Int,
  active: Boolean,
  sortOrder: Int
) extends APIMessage[ShopItem] {
  override def plan(connection: Connection): IO[Either[HttpError, ShopItem]] =
    PlanSteps.finish {
      for {
        existing <- EitherT(ShopCatalogSupport.requireItemRow(connection, itemId))
        timestamp = java.time.Instant.now().toString
        item <- EitherT(ShopCatalogSupport.requireUpdateItem(
          connection,
          existing.copy(
              name = name.trim,
              description = description.trim,
              price = price,
              currency = currency.trim,
              catalogIndex = catalogIndex,
              active = active,
              sortOrder = sortOrder,
              updatedAt = timestamp
            )
          )
        )
      } yield item
    }
}
